### ISSUE TITLE
Java grammar mixing abstractions

### ISSUE THREAD

**Issue by vsajip** [javacc21/javacc21#80]
With recent changes to the Java grammar, the generated code has things like this:
```java
public class StringLiteral extends Literal implements Expression {
```
I realise that this approach might be expedient, but I'm not sure it's a good idea to mix lexer-level abstractions with parser-level ones like this. I think the lexer level should be free of any parser-level abstractions. It might be better to have
```javacc
PrimaryExpression#void :
 (
  LiteralExpression
  |
 ...
```
and then
```javacc
LiteralExpression : Literal ;
```
and then `LiteralExpression` implements `Expression` as normal. It's not a purely abstract (no pun intended) concern - this plays into how the generated classes for Python would work. I had sort-of-working code for a Java parser with the old `Java.javacc` grammar, but now I get errors like this:
```
  File "[...]/javaparser/parser.py", line 7421, in parse_AssignmentExpression
    if (lhs.is_assignable_to()) and (self.next_token_type in self.first_setΣJava_javaccΣ521Σ5):
AttributeError: 'StringLiteral' object has no attribute 'is_assignable_to'
```
which alerted me to the issue.

What do you think about this?

**Comment by revusky** [comment]
Well, let's see... As you see, I did this refactoring where I got rid of `PrimaryExpression` as a node type. It exists as a production, but there is no `Node` called `PrimaryExpression`, since it is annotated with `#void`, right? I mean, the problem was that I wanted all these things to be a common type, in this case `Expression`, so earlier I had:

      INJECT PrimaryExpression : extends Expression

and we had:
   
     PrimaryExpression#

It was a *definite* node, so it was effectively a wrapper and anything like a StringLiteral would end up being "wrapped" inside of it. So in the generated AST, you'd always have: 

      PrimaryExpression
          StringLiteral

Right?

So this emerges from an idea that is well-motivated (I think...) that we could simplify the tree by just having no wrapper, just StringLiteral directly.

Well, regardless, since I set the goal of getting rid of the separate `PrimaryExpression` node, then you just directly have StringLiteral in the tree, but since it is highly convenient (at the very least) for all these things to have a common type, then I have the technical problem that this is only possible if `Expression` is an interface, not a base superclass.  By the way, I actually introduced the possibility of expressing this cleanly [here](https://javacc.com/2021/01/12/just-like-home-made/), that was 6 months ago, so I figured I would finally make use of this feature internally. But you see, meanwhile, I had various code here and there that would do things like: `args.childrenOfType(Expression.class)` to get the args list and without the intermediate `PrimaryExpression` wrapper, that (very natural) code would yield a spurious result, since it would ignore the StringLiteral and NumericalLiteral objects (which are directly Tokens) because they do not subclass `Expression`.

So, that, in very gory detail, is why I carried out the refactoring where `Expression` becomes an interface! And then the token subtypes can implement that interface, which is effectively a marker that indicates that these objects are expressions just on their own.

Now, as for the concern you're expressing about this new disposition "mixing abstractions", as best I can see it does not look like a big problem. In fact, to me offhand, it is arguably a good example of the principle of "information hiding". When you call a method like: `args.childrenOfType(Expression.class)` you get a `List<Expression>` back and you don't care whether the individual `Expression` objects are Tokens or inherit from BaseNode. (Or, in other words, I'm in the position that you are criticizing what I did (which is okay) but I actually thought I was doing something quite clever! It happens...) And you could be right.... 

But I have given this some thought. For example, [JavaLexer.javacc](https://github.com/javacc21/javacc21/blob/master/examples/java/JavaLexer.javacc) is INCLUDEd in [Java.javacc](https://github.com/javacc21/javacc21/blob/master/examples/java/Java.javacc#L55). But the various lines like: 

       INJECT StringLiteral : implements Expression

are in the `Java.javacc` file, NOT in `JavaLexer.javacc`, so if you were writing your own DSL but wanted to just re-use the JavaLexer part (because why create your own new lexical grammar when you can just reuse the Java one...) and you do the same INCLUDE instruction in your grammar, unless you specifically create an Expression node type and add the above INJECT, the StringLiteral token does not implement Expression. In fact, in an earlier pass, I added the injection to the `JavaLexer.javacc` file and then thought about it and realized that the file was no longer separately usable so I moved those injections to `Java.javacc`. 

Well, I guess I say all that to tell you that I am carefully thinking about this stuff. I am not convinced (though it is not impossible that you could convince me) that having things like StringLiteral and NumericalLiteral, which are directly tokens, implement an Expression interface is such a bad idea. Otherwise, you would have a kind of wrapper node, as you propose and have the wrapper node implement Expression -- which, actually was the case before. In the tree, you would have:

       PrimaryExpression
           StringLiteral

And that was really specifically because the wrapper could subclass `Expression` and a Token could not.

Well, this can all be revisited, of course, since it is such new stuff, but my initial thinking is that *as regards Java code* this new restructuring of things looks pretty okay. However, in terms of generating Python code, well... I can't speak to that and it may well cause a problem. Now, we do have the option of generating the wrapper `LiteralExpression` if generating Python and if not, then not.

     #if target==python
        LiteralExpression#
    #else
        LiteralExpression#void
   #endif
    

In other words, we could make it a definite node *when generating Python* and otherwise not. So, if this really was some kind of big issue *specifically* for Python, then we have the means to deal with it! (Of course, one can debate how elegant (or inelegant) tht is, I grant, but we can do it!) 

So hopefully, that answers your question...

**Comment by vsajip** [comment]
I've opened #86 which you will hopefully be willing to merge, after which I can close this issue.

**Comment by vsajip** [comment]
Closing, as the PR was merged (thanks!)
